File: backend/app/modules/group_entities_related_parties/compute.py
"""Derived model for Group Entities & Related Parties."""

from __future__ import annotations

from typing import Any

from app.modules.group_entities_related_parties.entities import (
    count_active_entities,
    count_entities_by_badge,
)
from app.modules.group_entities_related_parties.group_company import summarize_icdr_group_company_readiness
from app.modules.group_entities_related_parties.materiality import summarize_materiality_evaluation
from app.modules.group_entities_related_parties.ownership import (
    derive_ownership_chain_summary,
    summarize_ownership_paths,
)
from app.modules.group_entities_related_parties.rpt import calculate_rpt_summary


def compute_group_entities_model(
    payload: dict[str, Any],
    linked_references: dict[str, Any],
) -> dict[str, Any]:
    rpt_summary = calculate_rpt_summary(payload, linked_references)
    related_parties = (payload.get("relatedPartyUniverseAndClassification") or {}).get(
        "relatedPartyRelationships"
    ) or []

    classification = payload.get("groupCompanyAndMaterialityClassification") or {}
    icdr_determinations = [
        d for d in (classification.get("icdrGroupCompanyDeterminations") or []) if isinstance(d, dict)
    ]
    icdr_pending_board_count = sum(
        1 for d in icdr_determinations if d.get("classificationState") == "pending_board_determination"
    )
    icdr_group_company_count = sum(
        1
        for d in icdr_determinations
        if d.get("classificationState") in ("identified", "potentially_identified")
    )

    financial_readiness = (
        payload.get("groupEntityFinancialRegulatoryAndLitigationReadiness") or {}
    ).get("entityFinancialReadiness") or []

    ownership_summary = summarize_ownership_paths(payload)
    icdr_summary = summarize_icdr_group_company_readiness(payload)
    materiality_summary = summarize_materiality_evaluation(payload)

    return {
        "entityCount": count_active_entities(payload),
        "subsidiaryCount": count_entities_by_badge(payload, "subsidiary"),
        "stepDownSubsidiaryCount": count_entities_by_badge(payload, "step-down-subsidiary"),
        "associateCount": count_entities_by_badge(payload, "associate"),
        "jvCount": count_entities_by_badge(payload, "jv"),
        "promoterGroupEntityCount": count_entities_by_badge(payload, "promoter-group-entity"),
        "icdrGroupCompanyCount": icdr_group_company_count,
        "icdrPendingBoardCount": icdr_pending_board_count,
        "relatedPartyCount": sum(
            1
            for rp in related_parties
            if isinstance(rp, dict)
            and any(
                isinstance(fc, dict)
                and fc.get("currentHistorical") != "historical"
                and fc.get("related") == "yes"
                for fc in (rp.get("frameworkClassifications") or [])
            )
        ),
        "historicalRelatedPartyCount": sum(
            1
            for rp in related_parties
            if isinstance(rp, dict)
            and any(
                isinstance(fc, dict) and fc.get("currentHistorical") == "historical"
                for fc in (rp.get("frameworkClassifications") or [])
            )
        ),
        "ownershipRelationshipCount": len(
            (payload.get("ownershipControlAndRelationshipMapping") or {}).get(
                "ownershipRelationships"
            )
            or []
        ),
        "rptTransactionCount": len(
            (payload.get("relatedPartyTransactionsBalancesAndCommitments") or {}).get("transactions")
            or []
        ),
        "commonPursuitEntityCount": len(
            (payload.get("commonPursuitsDependenciesAndConflicts") or {}).get("commonPursuitRecords")
            or []
        ),
        "dependencyCount": len(
            (payload.get("commonPursuitsDependenciesAndConflicts") or {}).get(
                "interCompanyDependencies"
            )
            or []
        ),
        "negativeNetWorthCount": sum(
            1 for r in financial_readiness if isinstance(r, dict) and r.get("negativeNetWorth") == "yes"
        ),
        "lossMakingCount": sum(
            1 for r in financial_readiness if isinstance(r, dict) and r.get("lossMaking") == "yes"
        ),
        "auditorQualifiedCount": sum(
            1
            for r in financial_readiness
            if isinstance(r, dict) and r.get("auditorQualification") == "yes"
        ),
        "incompleteInformationCount": sum(
            1
            for r in financial_readiness
            if isinstance(r, dict)
            and r.get("informationStatus") in ("partial", "unavailable")
        ),
        "ibcConcernCount": sum(
            1
            for r in financial_readiness
            if isinstance(r, dict)
            and (
                r.get("ibcProceeding") == "yes"
                or r.get("windingUpPetition") == "yes"
                or r.get("liquidation") == "yes"
            )
        ),
        "pendingEntityInformationCount": sum(
            1
            for r in financial_readiness
            if isinstance(r, dict)
            and (
                r.get("informationStatus") == "not-requested"
                or r.get("informationReceived") == "no"
            )
        ),
        "rptSummary": rpt_summary,
        "ownershipChainSummary": derive_ownership_chain_summary(payload),
        "ownershipPathSummaries": ownership_summary,
        "icdrReadinessSummary": icdr_summary,
        "materialitySummary": materiality_summary,
    }

File: frontend/lib/group-entities-related-parties/compute.ts
/**
 * Derived model for Group Entities & Related Parties.
 */

import { countActiveEntities, countEntitiesByBadge, getEntities } from '@/lib/group-entities-related-parties/entities';
import { calculateRptSummary } from '@/lib/group-entities-related-parties/rpt';
import type { LinkedWorkstreamReferences } from '@/lib/group-entities-related-parties/types';
import type { GroupEntitiesRelatedPartiesPayload } from '@/lib/schemas/group-entities-related-parties';

export type GroupEntitiesModel = {
  entityCount: number;
  subsidiaryCount: number;
  stepDownSubsidiaryCount: number;
  associateCount: number;
  jvCount: number;
  promoterGroupEntityCount: number;
  icdrGroupCompanyCount: number;
  icdrPendingBoardCount: number;
  relatedPartyCount: number;
  historicalRelatedPartyCount: number;
  ownershipRelationshipCount: number;
  rptTransactionCount: number;
  commonPursuitEntityCount: number;
  dependencyCount: number;
  negativeNetWorthCount: number;
  lossMakingCount: number;
  auditorQualifiedCount: number;
  incompleteInformationCount: number;
  ibcConcernCount: number;
  pendingEntityInformationCount: number;
  rptSummary: ReturnType<typeof calculateRptSummary>;
};

export function computeGroupEntitiesModel(
  payload: GroupEntitiesRelatedPartiesPayload,
  linkedReferences: LinkedWorkstreamReferences,
): GroupEntitiesModel {
  const rptSummary = calculateRptSummary(payload, linkedReferences);
  const relatedParties = payload.relatedPartyUniverseAndClassification.relatedPartyRelationships;

  const icdrDeterminations =
    payload.groupCompanyAndMaterialityClassification.icdrGroupCompanyDeterminations;
  const icdrPendingBoardCount = icdrDeterminations.filter(
    (d) => d.classificationState === 'pending_board_determination',
  ).length;

  const icdrGroupCompanyCount = icdrDeterminations.filter(
    (d) => d.classificationState === 'identified' || d.classificationState === 'potentially_identified',
  ).length;

  const financialReadiness =
    payload.groupEntityFinancialRegulatoryAndLitigationReadiness.entityFinancialReadiness;

  return {
    entityCount: countActiveEntities(payload),
    subsidiaryCount: countEntitiesByBadge(payload, 'subsidiary'),
    stepDownSubsidiaryCount: countEntitiesByBadge(payload, 'step-down-subsidiary'),
    associateCount: countEntitiesByBadge(payload, 'associate'),
    jvCount: countEntitiesByBadge(payload, 'jv'),
    promoterGroupEntityCount: countEntitiesByBadge(payload, 'promoter-group-entity'),
    icdrGroupCompanyCount,
    icdrPendingBoardCount,
    relatedPartyCount: relatedParties.filter((rp) =>
      rp.frameworkClassifications.some((fc) => fc.currentHistorical !== 'historical' && fc.related === 'yes'),
    ).length,
    historicalRelatedPartyCount: relatedParties.filter((rp) =>
      rp.frameworkClassifications.some((fc) => fc.currentHistorical === 'historical'),
    ).length,
    ownershipRelationshipCount:
      payload.ownershipControlAndRelationshipMapping.ownershipRelationships.length,
    rptTransactionCount:
      payload.relatedPartyTransactionsBalancesAndCommitments.transactions.length,
    commonPursuitEntityCount:
      payload.commonPursuitsDependenciesAndConflicts.commonPursuitRecords.length,
    dependencyCount: payload.commonPursuitsDependenciesAndConflicts.interCompanyDependencies.length,
    negativeNetWorthCount: financialReadiness.filter((r) => r.negativeNetWorth === 'yes').length,
    lossMakingCount: financialReadiness.filter((r) => r.lossMaking === 'yes').length,
    auditorQualifiedCount: financialReadiness.filter((r) => r.auditorQualification === 'yes').length,
    incompleteInformationCount: financialReadiness.filter(
      (r) => r.informationStatus === 'partial' || r.informationStatus === 'unavailable',
    ).length,
    ibcConcernCount: financialReadiness.filter(
      (r) => r.ibcProceeding === 'yes' || r.windingUpPetition === 'yes' || r.liquidation === 'yes',
    ).length,
    pendingEntityInformationCount: financialReadiness.filter(
      (r) => r.informationStatus === 'not-requested' || r.informationReceived === 'no',
    ).length,
    rptSummary,
  };
}

export function deriveOwnershipChainSummary(payload: GroupEntitiesRelatedPartiesPayload): string[] {
  const lines: string[] = [];
  const entities = getEntities(payload);
  const issuer = entities.find((e) => e.classificationBadges.includes('parent')) ?? entities[0];
  if (!issuer) return lines;

  const relationships = payload.ownershipControlAndRelationshipMapping.ownershipRelationships.filter(
    (rel) => rel.currentHistorical !== 'historical',
  );

  for (const rel of relationships) {
    if (rel.investeeEntityId === issuer.id || rel.parentPartyEntityId === issuer.id) {
      const parent = entities.find((e) => e.id === rel.parentPartyEntityId);
      const child = entities.find((e) => e.id === rel.investeeEntityId);
      const parentName = parent?.identity.legalName || parent?.identity.displayName || rel.parentPartyEntityId.slice(0, 8);
      const childName = child?.identity.legalName || child?.identity.displayName || rel.investeeEntityId.slice(0, 8);
      lines.push(`${parentName} → ${childName} (${rel.relationshipType || 'relationship'})`);
    }
  }

  return lines.slice(0, 12);
}
